### PR TITLE
Support deployment of agent as a DaemonSet

### LINKED ISSUE #46: Support deployment of agent as a daemonset
Currently daemonset deployment strategy is not implemented for the jaeger-agent.

### PR DESCRIPTION
**Pull request by jpkrohling**
This PR adds support for deploying the agent as DaemonSet. 

Note that this PR includes an e2e test that makes use of the Docker image `jaegertracing/vertx-create-span:operator-e2e-tests`, whose source can be found here: https://gitlab.com/jpkroehling/vertx-create-span/blob/master/deployment.yaml . I'm not sure how to best proceed with those sources. A suggestion is to make a simple Go application that is similar to that vertx one and build it as part of this project's.

Closes #46 
Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

### CODE REVIEW

**Comment by jpkrohling** [comment]
This change is [Reviewable](https://reviewable.io/reviews/jaegertracing/jaeger-operator/52)

**Comment by objectiser** [comment]
@jpkrohling Would probably be a good idea to get someone with more Go experience to review.

In terms of the test - it would be good if such a test could be available in a more general location so that it could be shared - but if no obvious place at the moment then I guess it could be placed in this repo for now.

**Comment by jpkrohling** [comment]
> but if no obvious place at the moment then I guess it could be placed in this repo for now.

I'll create an issue to create such test here in this repo as a temporary measure, but I would like to do that in a follow-up PR.

**Comment by jpkrohling** [comment]
@objectiser said:
> LGTM, but would probably be a good idea to get someone with more Go experience to review.

@pavolloffay , would you like to review this one?

File: pkg/deployment/agent.go [hunk 1]
```diff
@@ -28,17 +30,94 @@ func NewAgent(jaeger *v1alpha1.Jaeger) *Agent {
...
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{{
+						Image: a.jaeger.Spec.Agent.Image,
+						Name:  "jaeger-agent-daemonset",
+						Args:  args,
+						Ports: []v1.ContainerPort{
+							{
+								ContainerPort: 5775,
+								HostPort:      5775,
+								Name:          "zk-compact-trft",
```
Comment (pavolloffay) [inline pkg/deployment/agent.go]: Is this populated to k8s objects? If so I would like to see the full names I had to think that means `zk-`

**Comment by pavolloffay** [comment]
the vertx app looks very simple I guess it only creates a span. We could have it here as a simple go service or even or job (no need for server).

File: pkg/deployment/agent.go (hunk 1 above)
Comment (jpkrohling) [reply pkg/deployment/agent.go]: It is, but unfortunately, there's a limit of 15 chars for this name. This is the best I could do with 15 chars while keeping it reasonably meaningful.

**Comment by jpkrohling** [comment]
> We could have it here as a simple go service or even or job (no need for server).

Will be done as part of #57 . For now, I just used what I already had.

File: pkg/deployment/agent.go (hunk 1 above)
Comment (pavolloffay) [reply pkg/deployment/agent.go]: that is weird IIRC in resource files we were using full names

Comment (jpkrohling) [reply pkg/deployment/agent.go]: For *Container* ports? Or for service names? For the `ContainerPort` object, this is what I see in [Kubernetes docs](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#containerport-v1-core):

> If specified, this must be an IANA_SVC_NAME and unique within the pod. Each named port in a pod must have a unique name. Name for the port that can be referred to by services.

Where IANA_SVC_NAME is:

> at most 15 characters, matching regex [a-z0-9]([a-z0-9-]*[a-z0-9])* and it must containts at least one letter [a-z], hypens cannot be adjacent to other hyphens): e.g. "http"

Perhaps you are

Comment (pavolloffay) [reply pkg/deployment/agent.go]: This https://github.com/jaegertracing/jaeger-kubernetes/blob/master/jaeger-production-template.yml#L78 it looks like a service port name

**Pull request merged**